src/work/Work.h

```diff
@@ -16,125 +17,94 @@ namespace stellar
-    // WORK_FAILURE_FATAL is equivalent to WORK_FAILURE_RAISE passed up in
-    // the work chain - when WORK_FAILURE_FATAL is raised in one work item,
-    // all work items that leaded to this one will also fail without retrying.
-    virtual State onSuccess();
-
-    static std::string stateName(State st);
-    State getState() const;
-    bool isDone() const;
-    void advance();
-    void reset();
+    void shutdown() override;
 
   protected:
-    std::weak_ptr<WorkParent> mParent;
-    std::string mUniqueName;
-    size_t mMaxRetries{RETRY_A_FEW};
-    size_t mRetries{0};
-    State mState{WORK_PENDING};
-    bool mScheduled{false};
-
-    std::unique_ptr<VirtualTimer> mRetryTimer;
-
-    std::function<void(asio::error_code const& ec)> callComplete();
-    void run();
-    void complete(CompleteResult result);
-    void scheduleComplete(CompleteResult result = WORK_COMPLETE_OK);
-    void scheduleRetry();
-    void scheduleRun();
-    void
-    scheduleSuccess()
-    {
-        scheduleComplete();
-    }
-    void
-    scheduleFailure()
-    {
-        scheduleComplete(WORK_COMPLETE_FAILURE);
-    }
-    void
-    scheduleFatalFailure()
+    Work(Application& app, std::string name,
+         size_t retries = BasicWork::RETRY_A_FEW);
+
+    // Note: `shared_from_this` assumes there exists a shared_ptr to the
+    // references class. This relates to who owns what in Work interface.
+    // Thus, `addWork` should be used to create Work (then the parent holds
+    // the reference).
+    template <typename T, typename... Args>
+    std::shared_ptr<T>
+    addWork(Args&&... args)
     {
-        scheduleComplete(WORK_COMPLETE_FATAL);
+        auto child = std::make_shared<T>(mApp, std::forward<Args>(args)...);
+        addChild(child);
+        child->startWork(wakeUpCallback());
+        wakeUp();
+        return child;
     }
 
-    void setState(State s);
+    State onRun() final;
+    void onReset() final;
 
-    void notifyParent();
-    virtual void notify(std::string const& childChanged) override;
+    // Implementers decide what they want to do: spawn more children,
+    // wait for all children to finish, or perform Work
+    virtual BasicWork::State doWork() = 0;
+
+    // Provide additional cleanup logic for reset
+    virtual void doReset();
 
   private:
-    VirtualClock::duration getRetryDelay() const;
+    std::list<std::shared_ptr<BasicWork>> mChildren;
+    std::list<std::shared_ptr<BasicWork>>::const_iterator mNextChild;
+
+    std::shared_ptr<BasicWork> yieldNextRunningChild();
+    void addChild(std::shared_ptr<BasicWork> child);
+    void clearChildren();
+
+    friend class WorkSequence;
+};
+
+/*
+ * WorkSequence is a helper class, that implementers can use if they
+ * wish to enforce the order of work execution. Users are required to pass
+ * a vector of BasicWork pointers in the expected execution order.
+ */
+class WorkSequence : public Work
+{
+    std::vector<std::shared_ptr<BasicWork>> mSequenceOfWork;
```

Comment:
Does that mean that WorkSequence has two lists of work: mSequenceOfWork and Work::mChildren? Seems like items from mSequenceOfWork are not shutted down when WorkSequence is.

Reply:
All items in mSequenceOfWork (except the one currently executing), are not part of the work tree (orchestrated by WorkScheduler) and are in reset/terminal state, and by definition cannot be running. However, it's probably good to shut them down as well for clarity, so that they are in proper DESTRUCTING state.